src/web/releases.rs

```diff
@@ -215,84 +216,84 @@ fn get_releases_by_owner(
+        name: String,
+        max_version: String,
+        description: Option<String>,
+        updated_at: DateTime<Utc>,
+    }
+    #[derive(Deserialize)]
+    struct CratesIoMeta {
+        total: u64,
     }
-    let offset = (page - 1) * limit;
-
-    let statement = "
-        SELECT
-            crates.name AS name,
-            releases.version AS version,
-            releases.description AS description,
-            releases.target_name AS target_name,
-            releases.release_time AS release_time,
-            releases.rustdoc_status AS rustdoc_status,
-            github_repos.stars AS github_stars,
-            COUNT(*) OVER() as total
-        FROM crates
-        INNER JOIN (
-            SELECT releases.id, releases.crate_id
-            FROM (
-                SELECT
-                    releases.id,
-                    releases.crate_id,
-                    RANK() OVER (PARTITION BY crate_id ORDER BY release_time DESC) as rank
-                FROM releases
-                WHERE releases.rustdoc_status AND NOT releases.yanked
-            ) AS releases
-            WHERE releases.rank = 1
-        ) AS latest_release ON latest_release.crate_id = crates.id
-        INNER JOIN releases ON latest_release.id = releases.id
-        LEFT JOIN github_repos ON releases.github_repo = github_repos.id
-        WHERE
-            ((char_length($1)::float - levenshtein(crates.name, $1)::float) / char_length($1)::float) >= 0.65
-            OR crates.name ILIKE CONCAT('%', $1, '%')
-        GROUP BY crates.id, releases.id, github_repos.stars
-        ORDER BY
-            levenshtein(crates.name, $1) ASC,
-            crates.name ILIKE CONCAT('%', $1, '%'),
-            releases.downloads DESC
-        LIMIT $2 OFFSET $3";
-
-    let rows = conn.query(statement, &[&query, &limit, &offset])?;
 
-    // Each row contains the total number of possible/valid results, just get it once
-    let total_results = rows
-        .get(0)
-        .map(|row| row.get::<_, i64>("total"))
-        .unwrap_or_default();
-    let packages: Vec<Release> = rows
+    use crate::utils::APP_USER_AGENT;
+    use once_cell::sync::Lazy;
+    use reqwest::blocking::Client as HttpClient;
+    use reqwest::header::{HeaderMap, HeaderValue, ACCEPT, USER_AGENT};
+
+    static HTTP_CLIENT: Lazy<HttpClient> = Lazy::new(|| {
+        let mut headers = HeaderMap::new();
+        headers.insert(USER_AGENT, HeaderValue::from_static(APP_USER_AGENT));
+        headers.insert(ACCEPT, HeaderValue::from_static("application/json"));
+        HttpClient::builder()
+            .default_headers(headers)
+            .build()
+            .unwrap()
+    });
+
+    let url = format!(
+        "https://crates.io/api/v1/crates?page={page}&per_page={limit}&q={query}",
+        page = page,
+        limit = limit,
+        query = query
+    );
+    debug!("fetching search results from {}", url);
+    let releases: CratesIoReleases = HTTP_CLIENT.get(&url).send()?.json()?;
```

Comment:
should all possible (error) status codes coming from the API lead to a server error here?

Reply:
I can imagine a few ways we could get an error here:

- If the API changes for some reason (??) - this shouldn't happen, cargo uses it and it's stable.
- We hit a rate limit.
- We send some sort of invalid query.

I'd consider all of those cases bugs and I think an internal error is right. I don't want to write error handling without knowing what error to handle.